pkg/engine/lifecycletest/target_test.go

```diff
@@ -1016,3 +1016,87 @@ func TestTargetedCreateDefaultProvider(t *testing.T) {
 	}
 	assert.True(t, foundDefaultProvider)
 }
+
+// TestEnsureUntargetedSame checks that an untargeted resource retains the prior state after an update when the provider
+// alters the inputs. This is a regression test for pulumi/pulumi#12964.
+func TestEnsureUntargetedSame(t *testing.T) {
+	t.Parallel()
+
+	// Provider that alters inputs during Check.
+	loaders := []*deploytest.ProviderLoader{
+		deploytest.NewProviderLoader("pkgA", semver.MustParse("1.0.0"), func() (plugin.Provider, error) {
+			return &deploytest.Provider{
+				CheckF: func(urn resource.URN,
+					olds, news resource.PropertyMap, _ []byte,
+				) (resource.PropertyMap, []plugin.CheckFailure, error) {
+					// Pulumi GCP provider alters inputs during Check.
+					news["__defaults"] = resource.NewStringProperty("exists")
+					return news, nil, nil
+				},
+			}, nil
+		}),
+	}
+
+	// Program that creates 2 resources.
+	program := deploytest.NewLanguageRuntime(func(_ plugin.RunInfo, monitor *deploytest.ResourceMonitor) error {
+		_, _, _, err := monitor.RegisterResource("pulumi:pulumi:Stack", "test-test", false)
+		assert.NoError(t, err)
+
+		_, _, _, err = monitor.RegisterResource("pkgA:m:typA", "resA", true, deploytest.ResourceOptions{
+			Inputs: resource.PropertyMap{
+				"foo": resource.NewStringProperty("foo"),
+			},
+		})
+		assert.NoError(t, err)
+
+		_, _, _, err = monitor.RegisterResource("pkgA:m:typA", "resB", true, deploytest.ResourceOptions{
+			Inputs: resource.PropertyMap{
+				"foo": resource.NewStringProperty("bar"),
+			},
+		})
+		assert.NoError(t, err)
+
+		return nil
+	})
+
+	host := deploytest.NewPluginHost(nil, nil, program, loaders...)
+	p := &TestPlan{}
+
+	project := p.GetProject()
+
+	// Set up stack with initial two resources.
+	origSnap, res := TestOp(Update).Run(project, p.GetTarget(t, nil), UpdateOptions{
+		Host: host,
+	}, false, p.BackendClient, nil)
+	assert.Nil(t, res)
+
+	// Target only `resA` and run a targeted update.
+	finalSnap, res := TestOp(Update).Run(project, p.GetTarget(t, origSnap), UpdateOptions{
+		Host: host,
+		UpdateTargets: deploy.NewUrnTargets([]string{
+			"urn:pulumi:test::test::pkgA:m:typA::resA",
+		}),
+	}, false, p.BackendClient, nil)
+	assert.Nil(t, res)
+
+	// Check that `resB` (untargeted) is the same between the two snapshots.
+	func() {
+		var initialState resource.State
+		for _, res := range origSnap.Resources {
+			if res.URN == "urn:pulumi:test::test::pkgA:m:typA::resB" {
+				initialState = *res
+			}
+		}
+		assert.NotNil(t, initialState, "initial `resB` state not found")
```

Comment:
This will always be true. `initialState` is a value, not a pointer.
You can drop the deref above and store the pointer.

```go
var initialState *resource.State
for ... {
  // ...
  initialState = res
}
```